pkg/steps/git_source_test.go

```diff
@@ -0,0 +1,107 @@
+package steps
+
+import (
+	"testing"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/util/diff"
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+)
+
+func TestDetermineRefsWorkdir(t *testing.T) {
+	testCases := []struct {
+		testName  string
+		refs      *prowapi.Refs
+		extraRefs []prowapi.Refs
+		expected  *prowapi.Refs
+	}{
+		{
+			testName:  "no workdir, nil Refs/ExtraRefs, expect nil",
+			refs:      nil,
+			extraRefs: nil,
+			expected:  nil,
+		},
+		{
+			testName:  "no workdir, expect Refs",
+			refs:      makeTestRefs(),
```

Comment:
We should prefer to have raw literal inputs and outputs to test cases unless they severely impede readability. Factoring out commonality and DRY in production code is useful as it ensures that there is one authoritative means by which some concept is represented. That does not apply to these tests -- there is nothing shared between the `refs` you choose to have in this test case versus the one after it. For all we care, they should be totally different. 

I would prefer to have test cases that are entirely self-contained within one `testCase`. As soon as I am needing to read helpers, it is harder to grok what a test case is doing and how and _why_ it shares logic with other tests.

Reply:
What you said make sense. For my defense, I didn't focus on the actual values of the `refs`, but I wanted to make sure which of them will be chosen. 
Either way, I will remove the helpers and have raw inputs and outputs since those cases are closer on what we should expect from the production.

Reply:
addressed